### PR TITLE
IGNITE-18510: Make composite publisher thread safe.

### PR DESCRIPTION
**Pull request by ygerzhedovich**

### CODE REVIEW

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/ConcatPublisher.java [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+import java.util.concurrent.Flow.Subscriber;
+import java.util.concurrent.Flow.Subscription;
+import java.util.concurrent.atomic.AtomicInteger;
+
+
+/**
+ * Thread-safe realization of combine multiple publishers. Generally, start consuming a source once the previous source has terminated,
+ * building a chain.
+ */
+public class ConcatPublisher<T> implements Publisher<T> {
```
Comment (AMashenkov) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/ConcatPublisher.java:34]:
```suggestion
public class ConcatenatedPublisher<T> implements Publisher<T> {
```

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/ConcatPublisher.java [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+        ConcatCoordinator<T> subscription = new ConcatCoordinator<>(downstream, sources);
+
+        downstream.onSubscribe(subscription);
+        subscription.drain();
+    }
+
+    /**
+     * Concatenation composite subscription.
+     */
+    static final class ConcatCoordinator<T> extends SubscriptionArbiter implements Subscriber<T> {
```
Comment (AMashenkov) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/ConcatPublisher.java:59]:
```suggestion
    static final class ConcatenatedSubscription<T> extends SubscriptionArbiter implements Subscriber<T> {
```

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/ConcatPublisher.java [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+                CURRENT = lk.findVarHandle(SubscriptionArbiter.class, "current", Subscription.class);
+                NEXT = lk.findVarHandle(SubscriptionArbiter.class, "next", Subscription.class);
+                DOWNSTREAM_REQUESTED = lk.findVarHandle(SubscriptionArbiter.class, "downstreamRequested", long.class);
+                PRODUCED = lk.findVarHandle(SubscriptionArbiter.class, "produced", long.class);
+                WIP = lk.findVarHandle(SubscriptionArbiter.class, "wip", int.class);
+
+            } catch (NoSuchFieldException | IllegalAccessException ex) {
+                throw new InternalError(ex);
+            }
+        }
```
Comment (AMashenkov) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/ConcatPublisher.java:154]:
```suggestion
        static final VarHandle CURRENT;
        static final VarHandle DOWNSTREAM_REQUESTED;
        static final VarHandle NEXT;
       static final VarHandle PRODUCED;
        static final VarHandle WIP;
        static {
            Lookup lk = MethodHandles.lookup();

            try {
                CURRENT = lk.findVarHandle(SubscriptionArbiter.class, "current", Subscription.class);
                NEXT = lk.findVarHandle(SubscriptionArbiter.class, "next", Subscription.class);
                DOWNSTREAM_REQUESTED = lk.findVarHandle(SubscriptionArbiter.class, "downstreamRequested", long.class);
                PRODUCED = lk.findVarHandle(SubscriptionArbiter.class, "produced", long.class);
                WIP = lk.findVarHandle(SubscriptionArbiter.class, "wip", int.class);

            } catch (NoSuchFieldException | IllegalAccessException ex) {
                throw new InternalError(ex);
            }
        }
        
        Subscription current;
                        
        Subscription next;

        long requested;

        long downstreamRequested;

        long produced;
 
        int wip;
```

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/ConcatPublisher.java [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+
+        ConcatCoordinator(
+                Subscriber downstream,
+                Iterator<Publisher<? extends T>> sources
+        ) {
+            this.downstream = downstream;
+            this.sources = sources;
+        }
+
+        @Override
```
Comment (AMashenkov) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/ConcatPublisher.java:77]: javadocs

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/OrderedMergePublisher.java [hunk 5]
```diff
@@ -36,49 +35,52 @@
  *
  * <p>Merges multiple concurrent ordered data streams into one.
  */
-public class SortingCompositePublisher<T> extends CompositePublisher<T> {
+public class OrderedMergePublisher<T> implements Publisher<T> {
```
Comment (AMashenkov) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/OrderedMergePublisher.java:38]:
```suggestion
/**
* <p>Merges multiple publishers using merge-sort algorithm. 
* NB: Upstream publishers must be sources of sorted data.
*/
public class MergedPublisher<T> implements Publisher<T> {
```

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/OrderedMergePublisher.java [hunk 6]
```diff
@@ -36,49 +35,52 @@
...
     }
 
     /**
      * Sorting composite subscription.
      *
      * <p>Merges multiple concurrent ordered data streams into one.
      */
-    public static class OrderedMergeCompositeSubscription<T> extends CompositePublisher.CompositeSubscription<T> {
-        /** Marker to indicate completed subscription. */
+    static final class OrderedMergeSubscription<T> implements Subscription {
```
Comment (AMashenkov) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/OrderedMergePublisher.java:79]:
```suggestion
    static final class MergedSubscription<T> implements Subscription {
```

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/OrderedMergePublisher.java [hunk 7]
```diff
@@ -107,14 +109,20 @@ public void subscribe(Subscriber<? super T> downstream) {
         /** Number of emitted rows (guarded by {@link #guardCntr}). */
         private long emitted;
 
+        static final VarHandle ERROR;
+
+        static final VarHandle CANCELLED;
+
+        static final VarHandle REQUESTED;
```
Comment (AMashenkov) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/OrderedMergePublisher.java:116]: Let's make static fields go first.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/ConcatPublisher.java [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+    }
+
+    /**
+     * Concatenation composite subscription.
+     */
+    static final class ConcatCoordinator<T> extends SubscriptionArbiter implements Subscriber<T> {
+        /** Counter to prevent concurrent execution of a critical section. */
+        private final AtomicInteger guardCntr = new AtomicInteger();
+
+        final Subscriber downstream;
```
Comment (xtern) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/ConcatPublisher.java:63]:
```suggestion
        final Subscriber<? super T> downstream;
```

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/ConcatPublisher.java [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+
+        final Iterator<Publisher<? extends T>> sources;
+
+        long consumed;
+
+        ConcatCoordinator(
+                Subscriber downstream,
```
Comment (xtern) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/ConcatPublisher.java:70]:
```suggestion
                Subscriber<? super T> downstream,
```

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/ConcatPublisher.java [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+                    }
+
+                    requested = req;
+
+                    if (nextSub != null && nextSub != this) {
+                        requestFrom = nextSub;
+                        requestAmount = req;
+                        CURRENT.compareAndSet(this, currentSub, nextSub);
+                    } else {
+                        // (6) --------------------------------------
```
Comment (xtern) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/ConcatPublisher.java:242]: do not forgot to remove

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/ConcatPublisher.java [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+        Subscription current;
+        static final VarHandle CURRENT;
+
+        Subscription next;
+        static final VarHandle NEXT;
+
+        long requested;
+
+        long downstreamRequested;
+        static final VarHandle DOWNSTREAM_REQUESTED;
```
Comment (xtern) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/ConcatPublisher.java:133]: I suggest to replace it with `AtomicLong`.
The code will be cleaner.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/ConcatPublisher.java [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+                    }
+
+                    sources.next().subscribe(this);
+
+                } while ((guardCntr.getAndDecrement() - 1) != 0);
+            }
+        }
+    }
+
+    static class SubscriptionArbiter implements Subscription {
```
Comment (xtern) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/ConcatPublisher.java:123]: We currently only have one class that inherits from `SubscriptionArbiter`.
It seems we should just combine `ConcatCoordinator` with `SubscriptionArbiter`

**Review by xtern — CHANGES_REQUESTED** [review]

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/ConcatPublisher.java (hunk 1 above)
Comment (xtern) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/ConcatPublisher.java]: Done

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/ConcatPublisher.java (hunk 3 above)
Comment (xtern) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/ConcatPublisher.java]: Done

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/OrderedMergePublisher.java (hunk 7 above)
Comment (xtern) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/OrderedMergePublisher.java:116]: Done

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/ConcatPublisher.java (hunk 4 above)
Comment (xtern) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/ConcatPublisher.java]: Done

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/ConcatPublisher.java (hunk 2 above)
Comment (xtern) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/ConcatPublisher.java]: Renamed to `ConcatenatedSubscriber`

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/OrderedMergePublisher.java (hunk 6 above)
Comment (xtern) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/OrderedMergePublisher.java:79]: It seems to me that the current names `OrderedMergePublisher` and `OrderedMergeSubscription` better reflect the purpose of these classes.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/ConcatPublisher.java [hunk 13]
```diff
@@ -0,0 +1,263 @@
...
+        long requested;
+
+        long downstreamRequested;
+        static final VarHandle DOWNSTREAM_REQUESTED;
+
+        long produced;
+        static final VarHandle PRODUCED;
+
+        int wip;
+        static final VarHandle WIP;
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/ConcatPublisher.java]: We decided to keep the code as close as possible to the original source https://akarnokd.blogspot.com/2017/09/java-9-flow-api-arbitration-and.html

**Review by xtern — APPROVED** [review]

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/OrderedMergePublisher.java (hunk 6 above)
Comment (AMashenkov) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/OrderedMergePublisher.java:81]: Concatenated publisher is also ordered, but not sorted, as it iterates over array of upstreams.
Another example, LinkedHashMap is ordered, but not sorted.

**Pull request merged**